protobuf/types/struct_test.go

```diff
@@ -0,0 +1,100 @@
+package types
+
+import (
+	"testing"
+
+	"github.com/mesg-foundation/engine/codec"
+	"github.com/mesg-foundation/engine/hash"
+	"github.com/stretchr/testify/require"
+)
+
+func TestStructHash(t *testing.T) {
+	hashes := make(map[string]bool)
+	structs := []Struct{
+		{Fields: map[string]*Value{}},
+		{Fields: map[string]*Value{"v": {Kind: &Value_NullValue{}}}},
+		{Fields: map[string]*Value{"v": {Kind: &Value_NumberValue{}}}},
+		{Fields: map[string]*Value{"v": {Kind: &Value_StringValue{}}}},
+		{Fields: map[string]*Value{"v": {Kind: &Value_BoolValue{}}}},
+		{Fields: map[string]*Value{"v": {Kind: &Value_StructValue{}}}},
+		{Fields: map[string]*Value{"v": {Kind: &Value_ListValue{}}}},
+	}
+	for _, s := range structs {
+		hashes[string(hash.Dump(s))] = true
+	}
+	require.Equal(t, len(hashes), len(structs))
+}
+
+func TestStructMarshal(t *testing.T) {
+	var (
+		structSort1 = &Struct{
+			Fields: map[string]*Value{
+				"10000":  {Kind: &Value_NumberValue{NumberValue: 10}},
+				"1":      {Kind: &Value_StringValue{StringValue: "valuea"}},
+				"null":   {Kind: &Value_NullValue{NullValue: NullValue_NULL_VALUE}},
+				"number": {Kind: &Value_NumberValue{NumberValue: 10}},
+				"string": {Kind: &Value_StringValue{StringValue: "valuea"}},
+				"bool":   {Kind: &Value_BoolValue{BoolValue: true}},
+				"struct": {Kind: &Value_StructValue{StructValue: &Struct{Fields: map[string]*Value{
+					"10000":  {Kind: &Value_NumberValue{NumberValue: 10}},
+					"1":      {Kind: &Value_StringValue{StringValue: "valuea"}},
+					"null":   {Kind: &Value_NullValue{NullValue: NullValue_NULL_VALUE}},
+					"number": {Kind: &Value_NumberValue{NumberValue: 10}},
+					"struct": {Kind: &Value_StructValue{StructValue: &Struct{Fields: map[string]*Value{
+						"10000":  {Kind: &Value_NumberValue{NumberValue: 10}},
+						"1":      {Kind: &Value_StringValue{StringValue: "valuea"}},
+						"null":   {Kind: &Value_NullValue{NullValue: NullValue_NULL_VALUE}},
+						"number": {Kind: &Value_NumberValue{NumberValue: 10}},
+					}}}},
+				}}}},
+				"list": {Kind: &Value_ListValue{ListValue: &ListValue{Values: []*Value{
+					{Kind: &Value_StringValue{StringValue: "dvaluea"}},
+				}}}},
```

Comment:
todo?

- s/valuea/value/
- change some keys (for example I'm not sure why `StringValue` has key `1` - maybe string_1?)
- Key10000 hodls value 10 - s/10000/10/ ?

Reply:
The mess with the keys is to try to create the most random ordered map to test if the sort is correctly working.
I find out that for some reason, if the keys are only one alpha characters, then the map is "automatically" sorted. That's why some keys are numeric and some other are just the type of value